### ISSUE TITLE
Consider `execv` for `uv run` on Unix

### ISSUE THREAD

**Issue by zanieb** [astral-sh/uv#3095]
See https://github.com/astral-sh/uv/pull/3074#issuecomment-2060847595

**Comment by bluss** [comment]
For the https://github.com/bluss/pyproject-local-kernel usecase (sorry if this is boring), spawn vs exec would make a difference on Linux in terms of how a jupyter notebook is kernel is launched.

jupyter lab launches my python script -> launches uv run -> launches ipykernel.

My script already handles forwarding posix signals to the child (first arrow). If you use exec, the second arrow disappears and posix signals are forwarded to the ipykernel which is useful. I think exec is generally what you'd want to do. I didn't run into this with rye run, because it uses exec on Linux.

The kernel [can be configured with interrupt_mode=message](https://jupyter-client.readthedocs.io/en/latest/kernels.html#kernel-specs) (instead of default signal) to make this issue disappear, but it's still interesting.

**Comment by bluss** [comment]
Is this still possible, how does the ephemeral venv cleanup work? Is it using a scope based cleanup. I'll file another issue that's related to ipykernel, and maybe we'll find a good solution.

**Comment by zanieb** [comment]
It is not clear how we'd clean up the ephemeral environment, that's a good question.

**Comment by charliermarsh** [comment]
It runs on Drop.

**Comment by bluss** [comment]
The pyproject-local-kernel usecase seems to work perfectly on Linux now after the #5257 fix I think :star_struck:. Probably should carry over to windows too. Thanks for the help with this :slightly_smiling_face:

uv run --with ipykernel is perfect for the notebook use case, it's very smooth.

**Comment by charliermarsh** [comment]
Awesome. I did test on my Windows machine and it worked as expected for me, so hopefully good there too...

**Comment by zanieb** [comment]
When you `execv` is nothing dropped until the process exits? I'm quite naive as to how that is handled. cc @BurntSushi

**Comment by BurntSushi** [comment]
Basically, you can think of `execv` (and related functions) as like a `std::process::exit`. It can be inserted anywhere in your program, and it will unceremoniously stop executing the current program. Destructors are not guaranteed to run.

So to make this work here, we'd probably want to bubble some instruction to "run this `exec` command" to `main` in a way that everything else goes out of scope and gets dropped. Otherwise, you'll not just have things like an ephemeral environment hanging around, but all the various allocations and file descriptors won't get dropped either (they will eventually by the OS once the replacing process terminates, AIUI).

Also, we are very likely already using `exec`. I think what is meant here is that we do not want to _fork_ before an exec. Or stated at a higher level, we want to replace the current process with the thing we want to run.

**Comment by zanieb** [comment]
Hm.. thanks for the details.

I think that supports my concern that we have resources e.g., the ephemeral environment, that can't be removed until the spawned process finishes which makes it hard to entirely cede control?

**Comment by charliermarsh** [comment]
Yeah I think the general form of the problem here is that `execv` won't run Drop, but it we Drop things before deferring to `execv`, we'll remove things we need like temporary directories.

**Comment by zanieb** [comment]
This feels like a pretty significant blocker for switching to `execv` – I think we might need to prioritize passing signals to the child in a robust manner?

**Comment by charliermarsh** [comment]
Yeah, agreed.

**Comment by zanieb** [comment]
Signal handling was improved https://github.com/astral-sh/uv/pull/11009

I think the `Drop` question remains a blocker. I'll leave this open for future consideration though.

**Comment by zanieb** [comment]
More signal handling changes in https://github.com/astral-sh/uv/pull/13017

**Comment by tapetersen** [comment]
While signal propagation to the child correctly solves many problems one issue it still creates is that it blocks systemd's ready-notification in default configuration as it ignores messages from any other process than the root process.
That can be worked around by also using `NotifyAccess=all` see  https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#NotifyAccess= 

That also has it's downsides though.

Mostly adding as another usecase/reason where child process isn't really the same.

I still don't have a good Idea how handling cleanup in the general case of using exec instead of spawn would work but for the systemd-case it of course has it's own way of providing ethemereal or persistant directories for services which ideally would be used here as well (passed through env-vars as `$RUNTIMEDIR` and `$CACHEDIR` for those that would be relevant here).

In those cases something that can just set up the environment needed  somewhere specific and not care itself about would be useful but I'm not sure if it's useful enought that uv should include it.
